src/elements/VectorPotential.H

```diff
@@ -0,0 +1,444 @@
+
+            particle2(1) = x2 + tau * dHdpx;
+            particle2(2) = px2;
+            particle2(3) = y2 + tau * dHdpy;
+            particle2(4) = py2;
+            particle2(5) = t2 + tau * dHdpt;
+            particle2(6) = pt2;
+
+            zeval += tau;
+        }
+
+
+        /** This evaluates the derivatives of the Hamiltonian at a point.
+         *
+         * @param x horizontal coordinate in m
+         * @param px horizontal momentum
+         * @param y vertical coordinate in m
+         * @param py vertical momentum
+         * @param t temporal coordinate in m
+         * @param pt temporal momentum (energy deviation)
+         * @param zeval lattice position in m
+         * @param[out] tuple containing derivatives of the Hamiltonian
+         */
+        std::tuple <amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal,
+               amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal>
+        Hamiltonian_Gradient (amrex::ParticleReal x, amrex::ParticleReal px,
+                   amrex::ParticleReal y, amrex::ParticleReal py,
+                   [[maybe_unused]] amrex::ParticleReal t, amrex::ParticleReal pt,
+                   [[maybe_unused]] amrex::ParticleReal zeval) const
+        {
+            using namespace amrex::literals; // for _rt and _prt
+
+            amrex::ParticleReal dHdx = 0_prt;
+            amrex::ParticleReal dHdpx = 0_prt;
+            amrex::ParticleReal dHdy = 0_prt;
+            amrex::ParticleReal dHdpy = 0_prt;
+            amrex::ParticleReal dHdt = 0_prt;
+            amrex::ParticleReal dHdpt = 0_prt;
+
+            // Evaluate the vector potential and its derivatives
+            auto [ax,ay,daxdx,daxdy,daydx,daydy,dazdx,dazdy] = vector_potential_eval(x,y,t,zeval);
+
+            // Implementation of derivatives for rectilinear magnetostatic element (static units):
+            // Radical appearing in the Hamiltonian (equal to the value of pz):
+            amrex::ParticleReal const pxmech = px - ax;
+            amrex::ParticleReal const pymech = py - ay;
+            amrex::ParticleReal const radical = std::sqrt(1_prt + pt*pt - px*px - py*py - 2_prt*pt*m_ibeta);
+
+            dHdx = - (pxmech * daxdx + pymech * daydx) / radical - dazdx;
+            dHdy = - (pxmech * daxdy + pymech * daydy) / radical - dazdy;
+            dHdt = 0_prt;
+
+            dHdpx = pxmech / radical;
+            dHdpy = pymech / radical;
+            dHdpt = - m_ibeta + (m_ibeta - pt) / radical;
+
+            return std::make_tuple(dHdx,dHdpx,dHdy,dHdpy,dHdt,dHdpt);
+
+         }
+
+
+        /** This evaluates the vector potential and derivatives at a point.
+         *
+         * @param x horizontal coordinate in m
+         * @param y vertical coordinate in m
+         * @param t temporal coordinate in m
+         * @param zeval lattice position in m
+         * @param[out] tuple containing the vector potential and derivatives
+         */
```

Comment:
This function is hard-coded at the moment, to represent the test case of a quad.  These quantities we would like to obtain directly from the user using the math parser.

Reply:
Thanks for the annotation! Updated :)